File: servers/get_env_info_server.py
"""MCP server — expose environment information tools.

Tools:
    get_system_info   — host OS info
    get_cpu_info      — processor / cores
    get_memory_info   — physical memory
    get_disk_info     — C: drive usage
    get_gpu_info      — GPU adapters
    get_runtime_info  — installed dev tools & runtimes
"""

import asyncio
import os
import sys

from mcp.server import Server
from mcp.server.stdio import stdio_server
from mcp.types import TextContent, Tool

try:
    from tools import load_all
except ImportError:
    _proj_root = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
    sys.path.insert(0, _proj_root)
    from tools import load_all

_ENV_TOOLS = frozenset(
    [
        "get_system_info",
        "get_cpu_info",
        "get_memory_info",
        "get_disk_info",
        "get_gpu_info",
        "get_runtime_info",
    ]
)

server = Server("get-env-info")


@server.list_tools()
async def list_tools() -> list[Tool]:
    tools = load_all()
    return [
        Tool(
            name=t.name,
            description=t.description,
            inputSchema=t.parameters,
        )
        for t in tools.values()
        if t.name in _ENV_TOOLS
    ]


@server.call_tool()
async def call_tool(name: str, arguments: dict) -> list[TextContent]:
    """Execute a tool without blocking the event loop.

    Long-running synchronous tools (e.g. ``get_runtime_info`` which scans
    PATH for 150+ executables) are offloaded to a thread so the MCP
    server can keep processing messages.
    """
    tools = load_all()
    if name not in tools:
        return [TextContent(type="text", text=f"Error: unknown tool '{name}'")]
    try:
        result = await asyncio.to_thread(tools[name], **arguments)
        return [TextContent(type="text", text=result)]
    except Exception as e:
        return [TextContent(type="text", text=f"Error: {e}")]


async def main() -> None:
    async with stdio_server() as (read_stream, write_stream):
        await server.run(
            read_stream,
            write_stream,
            server.create_initialization_options(),
        )


if __name__ == "__main__":
    asyncio.run(main())

File: servers/get_env_info_server.ts
/**
 * MCP server — expose environment information tools.
 *
 * Tools:
 *   get_system_info   — host OS info
 *   get_cpu_info      — processor / cores
 *   get_memory_info   — physical memory
 *   get_disk_info     — C: drive usage
 *   get_gpu_info      — GPU adapters
 *   get_runtime_info  — installed dev tools & runtimes
 */

import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { StdioServerTransport } from "@modelcontextprotocol/sdk/server/stdio.js";
import { createRequire } from "module";
import { getAllTools } from "../tools/index.js";

const ENV_TOOL_NAMES = new Set([
  "get_system_info",
  "get_cpu_info",
  "get_memory_info",
  "get_disk_info",
  "get_gpu_info",
  "get_runtime_info",
]);

// version 单一来源：根 package.json（createRequire 规避 ESM JSON import 兼容性）
const pkg = createRequire(import.meta.url)("../package.json");

const server = new McpServer({
  name: "get-env-info",
  version: pkg.version,
});

// Register each environment tool with the MCP server
try {
  for (const [name, tool] of getAllTools()) {
    if (ENV_TOOL_NAMES.has(name)) {
      server.tool(
        name,
        tool.description,
        tool.parameters,
        { readOnlyHint: true, destructiveHint: false, idempotentHint: true, openWorldHint: false },
        async (args: Record<string, unknown>) => {
          try {
            const result = await tool.handler(args);
            return { content: [{ type: "text" as const, text: result }] };
          } catch (e) {
            return {
              content: [{ type: "text" as const, text: `Error: ${e}` }],
            };
          }
        }
      );
    }
  }

  const transport = new StdioServerTransport();
  await server.connect(transport);
} catch (e) {
  console.error(`[get-env-info] Failed to start: ${e}`);
  process.exit(1);
}
